File: NCoreUtils.AspNetCore.Rest/HttpResponseOutput.fs
namespace NCoreUtils.AspNetCore.Rest

open System
open System.IO
open Microsoft.AspNetCore.Http


type internal HttpResponseOutput (response : HttpResponse) =
  interface IConfigurableOutput<Stream> with
    member __.AsyncInitialize info =
      if info.Length.HasValue then
        response.ContentLength <- info.Length
      if not (String.IsNullOrEmpty info.ContentType) then
        response.ContentType <- info.ContentType
      async.Return response.Body

File: NCoreUtils.AspNetCore.Rest/Rest/Internal/HttpResponseOutput.cs
using System;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;

namespace NCoreUtils.AspNetCore.Rest.Internal;

internal sealed class HttpResponseOutput(HttpResponse response) : IConfigurableOutput<Stream>
{
    private readonly HttpResponse _response = response ?? throw new ArgumentNullException(nameof(response));

    public ValueTask<Stream> InitializeAsync(OutputInfo info, CancellationToken cancellationToken)
    {
        if (info.Length.HasValue)
        {
            _response.ContentLength = info.Length.Value;
        }
        if (!string.IsNullOrEmpty(info.ContentType))
        {
            _response.ContentType = info.ContentType;
        }
        return new ValueTask<Stream>(_response.Body);
    }
}